schema broken (GraphQLLocatedError for 'state')

Attempting some development today I found that the end-to-end system has broken so that using the current ``master`` on all three repositories ``cylc-flow``, ``cylc-ui`` & ``cylc-uiserver`` means that (at least for the several suites I tested on), the UI Tree View gives a blank page apart from the top bar with the suite name (see the screenshot below), & further digging uncovers some error messages.

I have traced this via reverting commits (on each of the three repos independently until I hit the case where the full system went from working as previous to hitting the above problem) to #3211, which unfortunately has broken the schema, throwing up errors relating to 'state', as below.

I'm going to trace back the logic to work out which aspect(s) of that PR (with one commit) are to blame, & put in a PR to fix them. Thanks.

### Error messages

#### GraphiQL

From interactive GraphQL, I got the following error response returned:

```graphql
  "errors": [
    {
      "message": "state",
      "locations": [
        {
          "line": 21,
          "column": 5
        }
      ],
      "path": [
        "workflows",
        0,
        "tasks"
      ]
    }
  ],
```
(if relevant, this was from a basic query on one of my running workflows:)
```graphql
query workflows($tIds: [ID]){
  workflows(ids: "subsuite-app-domingo"){
    id
    name
    status
    stateTotals {
      runahead
      waiting
      held
      queued
      expired
      ready
      submitFailed
      submitRetrying
      submitted
      retrying
      running
      failed
      succeeded
    }
    tasks(ids: $tIds){
      name
      meta {
        title
        description
        URL
        userDefined
      }
      proxies(ids: $tIds){
        cyclePoint
        state
      }
    }
  }
}
```

### Terminal

And a traceback to the terminal was consistent with it being an issue on ``state``, emerging from the ``schema.py``.
```
ERROR:graphql.execution.utils:Traceback (most recent call last):
  File "/home/h06/sbarth/miniconda3/envs/py3env/lib/python3.7/site-packages/promise/promise.py", line 842, in handle_future_result
    resolve(future.result())
  File "/net/home/h06/sbarth/cylc.git/cylc/flow/network/schema.py", line 263, in get_nodes_by_id
    return await resolvers.get_nodes_by_id(node_type, args)
  File "/net/home/h06/sbarth/cylc-uiserver/resolvers.py", line 65, in get_nodes_by_id
    return [node for node in nodes
  File "/net/home/h06/sbarth/cylc-uiserver/resolvers.py", line 67, in <listcomp>
    node_filter(node, args))]
  File "/net/home/h06/sbarth/cylc.git/cylc/flow/network/resolvers.py", line 90, in node_filter
    (args.get('ghosts') or node.state != '') and
graphql.error.located_error.GraphQLLocatedError: state
```

**************************

**Release version(s) and/or repository branch(es) affected?**
``8.0a0``.

**Steps to reproduce the bug**

* Update all three repos to the current ``master`` branches.
* Start up the UI Server & open the UI, as usual.
* Run some suites.
* Open them in Tree View, & see that the page is there but the table is missing, with no data there emerging.

**Expected behavior**

* A Tree View (nested) table populated with its task data, for each suite.

**Screenshots**

| Current ``master`` | After reverting commit from #3211 |
|---|---|
| ![Screenshot from 2019-07-13 19-02-04](https://user-images.githubusercontent.com/30274190/61175052-20caf380-a5a1-11e9-856f-51444f4b77e9.png)  | ![Screenshot from 2019-07-13 19-06-10](https://user-images.githubusercontent.com/30274190/61175078-68517f80-a5a1-11e9-8b2e-8445a76880a6.png) |

`master` giving the error for the query.

![image](https://user-images.githubusercontent.com/304786/61208729-2b75bc00-a74c-11e9-89ad-0f06f90cc87e.png)


Query worked on this branch, with no errors in the console.

![image](https://user-images.githubusercontent.com/304786/61208997-b6ef4d00-a74c-11e9-9101-d14da00de13f.png)

One strange failure in the unit tests, but not sure how that can be related.

```
=================================== FAILURES ===================================
___________________ TestHostAppointer.test_remove_bad_hosts ____________________
self = <flow.tests.test_host_appointer.TestHostAppointer testMethod=test_remove_bad_hosts>
    def test_remove_bad_hosts(self):
        """Test the '_remove_bad_hosts' method.
        Test using 'localhost' only since remote host functionality is
        contained only inside remote_cylc_cmd() so is outside of the scope
        of HostAppointer.
        """
        self.mock_global_config(set_hosts=['localhost'])
>       self.assertTrue(self.app._remove_bad_hosts().get('localhost', False))
E       AssertionError: False is not true
cylc/flow/tests/test_host_appointer.py:233: AssertionError
------------------------------ Captured log call -------------------------------
host_appointer.py          200 WARNING  can't get host metric from 'localhost'get-host-metrics  # returncode=1, err=Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (graphql-core 2.2 (/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages), Requirement.parse('graphql-core<2,>=0.5.0'), {'graphql-relay'})
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.7.1/bin/cylc-get-host-metrics", line 4, in <module>
    __import__('pkg_resources').require('cylc-flow==8.0a0')
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3191, in <module>
    @_call_aside
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3175, in _call_aside
    f(*args, **kwargs)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 3204, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 585, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 598, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (graphql-core 2.2 (/home/travis/virtualenv/python3.7.1/lib/python3.7/site-packages), Requirement.parse('graphql-core<2,>=0.5.0'), {'graphql-relay'})
```

--- a/cylc/flow/network/resolvers.py
+++ b/cylc/flow/network/resolvers.py
@@ -21,7 +21,9 @@ from fnmatch import fnmatchcase
 
 # Message Filters
 def collate_workflow_atts(workflow):
-    """Collate workflow filter attributes."""
+    """Collate workflow filter attributes, setting defaults if non-existent."""
+    # Append new atts to the end of the list,
+    # this will retain order used in index access.
     return [
         workflow.owner,
         workflow.name,
@@ -29,27 +31,32 @@ def collate_workflow_atts(workflow):
     ]
 
 
-def workflow_atts_filter(natts, items):
-    """Match components of id argument with those of workflow id."""
+def workflow_ids_filter(w_atts, items):
+    """Match id arguments with workflow attributes.
+
+    Returns a boolean."""
+    # Return true if workflow matches any id arg.
     for owner, name, status in set(items):
-        if ((not owner or fnmatchcase(natts[0], owner)) and
-                (not name or fnmatchcase(natts[1], name)) and
-                (not status or natts[2] == status)):
+        if ((not owner or fnmatchcase(w_atts[0], owner)) and
+                (not name or fnmatchcase(w_atts[1], name)) and
+                (not status or w_atts[2] == status)):
             return True
     return False
 
 
 def workflow_filter(flow, args):
     """Filter workflows based on attribute arguments"""
-    natts = collate_workflow_atts(flow.workflow)
+    w_atts = collate_workflow_atts(flow.workflow)
+    # The w_atts (workflow attributes) list contains ordered workflow values
+    # or defaults (see collate function for index item).
     return ((not args.get('workflows') or
-            workflow_atts_filter(natts, args['workflows'])) and
+            workflow_ids_filter(w_atts, args['workflows'])) and
             not (args.get('exworkflows') and
-            workflow_atts_filter(natts, args['exworkflows'])))
+            workflow_ids_filter(w_atts, args['exworkflows'])))
 
 
 def collate_node_atts(node):
-    """Collate node filter attributes."""
+    """Collate node filter attributes, setting defaults if non-existent."""
     node_id = getattr(node, 'id')
     slash_count = node_id.count('/')
     n_id = node_id.split('/', slash_count)
@@ -59,6 +66,14 @@ def collate_node_atts(node):
     else:
         n_cycle = n_id[2]
         n_name = n_id[3]
+    # Append new atts to the end of the list,
+    # this will retain order used in index access
+    # 0 - owner
+    # 1 - workflow
+    # 2 - Cycle point or None
+    # 3 - name or namespace list
+    # 4 - submit number or None
+    # 5 - state
     return [
         n_id[0],
         n_id[1],
@@ -69,33 +84,38 @@ def collate_node_atts(node):
     ]
 
 
-def node_atts_filter(natts, items):
-    """Match node id argument with node attributes."""
+def node_ids_filter(n_atts, items):
+    """Match id arguments with node attributes.
+
+    Returns a boolean."""
     for owner, workflow, cycle, name, submit_num, state in items:
-        if ((not owner or fnmatchcase(natts[0], owner)) and
-                (not workflow or fnmatchcase(natts[1], workflow)) and
-                (not cycle or fnmatchcase(natts[2], cycle)) and
-                any(fnmatchcase(nn, name) for nn in natts[3]) and
+        if ((not owner or fnmatchcase(n_atts[0], owner)) and
+                (not workflow or fnmatchcase(n_atts[1], workflow)) and
+                (not cycle or fnmatchcase(n_atts[2], cycle)) and
+                any(fnmatchcase(nn, name) for nn in n_atts[3]) and
                 (not submit_num or
-                    fnmatchcase(str(natts[4]), submit_num.lstrip('0'))) and
-                (not state or natts[5] == state)):
+                    fnmatchcase(str(n_atts[4]), submit_num.lstrip('0'))) and
+                (not state or n_atts[5] == state)):
             return True
     return False
 
 
 def node_filter(node, args):
     """Filter nodes based on attribute arguments"""
-    natts = collate_node_atts(node)
+    n_atts = collate_node_atts(node)
+    # The n_atts (node attributes) list contains ordered node values
+    # or defaults (see collate function for index item).
     return (
-        (args.get('ghosts') or node.state != '') and
-        (not (args.get('states') and node.state != '')
-         or node.state in args['states']) and
-        not (args.get('exstates') and node.state != '' and
-             node.state in args['exstates']) and
+        (args.get('ghosts') or n_atts[5] != '') and
+        (not (args.get('states') and n_atts[5] != '')
+         or n_atts[5] in args['states']) and
+        not (args.get('exstates') and n_atts[5] != '' and
+             n_atts[5] in args['exstates']) and
         (args.get('mindepth', -1) < 0 or node.depth >= args['mindepth']) and
         (args.get('maxdepth', -1) < 0 or node.depth <= args['maxdepth']) and
-        (not args.get('ids') or node_atts_filter(natts, args['ids'])) and
-        not (args.get('exids') and node_atts_filter(natts, args['exids']))
+        # Now filter node against id arg lists
+        (not args.get('ids') or node_ids_filter(n_atts, args['ids'])) and
+        not (args.get('exids') and node_ids_filter(n_atts, args['exids']))
     )
 
 

